config/translate/translate.go

```diff
@@ -0,0 +1,202 @@
+		return true
+	case isInvalidInConfig(k1):
+		panic("Encountered invalid kind in config. This is a bug, please file a report")
+	case k1 == reflect.Ptr || k1 == reflect.Slice:
+		return sameType(t1.Elem(), t2.Elem())
+	case k1 == reflect.Struct:
+		return sameStruct(t1, t2)
+	default:
+		panic("Encountered unknown kind in config. This is a bug, please file a report")
+	}
+}
+
+// precondition: t1, t2 are both of Kind 'struct'
+func sameStruct(t1, t2 reflect.Type) bool {
+	if t1.NumField() != t2.NumField() || t1.Name() != t2.Name() {
+		return false
+	}
+	for i := 0; i < t1.NumField(); i++ {
+		t1f := t1.Field(i)
+		t2f, ok := t2.FieldByName(t1f.Name)
+
+		if !ok || !sameType(t1f.Type, t2f.Type) {
+			return false
+		}
+	}
+	return true
+}
+
+// checks that t could reasonably be the type of a translator function
+func probablyValidTranslator(t reflect.Type) bool {
+	if t.Kind() != reflect.Func {
+		return false
+	}
+	if t.NumIn() != 1 || t.NumOut() != 1 {
+		return false
+	}
+	if isInvalidInConfig(t.In(0).Kind()) || isInvalidInConfig(t.Out(0).Kind()) {
+		return false
+	}
+	return true
+}
+
+// translate from one type to another, but deep copy all data
+// precondition: vFrom and vTo are the same type as defined by sameType()
+// precondition: vTo is addressable and settable
+func (t translator) translateSameType(vFrom, vTo reflect.Value) {
+	k := vFrom.Kind()
+	switch {
+	case isPrimitive(k):
+		// Use convert, even if not needed; type alias to primitives are not
+		// directly assignable and calling Convert on primitives does no harm
+		vTo.Set(vFrom.Convert(vTo.Type()))
+	case k == reflect.Ptr:
+		if vFrom.IsNil() {
+			return
+		}
+		vTo.Set(reflect.New(vTo.Type().Elem()))
+		t.translate(vFrom.Elem(), vTo.Elem())
+	case k == reflect.Slice:
+		if vFrom.IsNil() {
+			return
+		}
+		vTo.Set(reflect.MakeSlice(vTo.Type(), vFrom.Len(), vFrom.Len()))
+		for i := 0; i < vFrom.Len(); i++ {
+			t.translate(vFrom.Index(i), vTo.Index(i))
+		}
+	case k == reflect.Struct:
+		for i := 0; i < vFrom.NumField(); i++ {
+			t.translate(vFrom.Field(i), vTo.FieldByName(vFrom.Type().Field(i).Name))
+		}
+	default:
+		panic("Encountered types that are not the same when they should be. This is a bug, please file a report")
+	}
+}
+
+// v2 must be addressable, should be acquired by calling reflect.ValueOf() on a variable of the correct type
+func (t translator) translate(vFrom, vTo reflect.Value) {
+	tFrom := vFrom.Type()
+	tTo := vTo.Type()
+	if fnv := t.getTranslator(tFrom, tTo); fnv.IsValid() {
+		vTo.Set(fnv.Call([]reflect.Value{vFrom})[0])
+		return
+	}
+	if sameType(tFrom, tTo) {
+		t.translateSameType(vFrom, vTo)
+		return
+	}
+
+	panic(fmt.Sprintf("Translator not defined for %v to %v", tFrom, tTo))
+}
+
+type Translator interface {
+	AddCustomTranslator(t interface{})
+	Translate(from, to interface{})
+}
+
+func NewTranslator() Translator {
+	return &translator{}
+}
+
+type translator struct {
+	// List of translation funcs, must pass probablyValidTranslator
```

Comment:
I'd call this "List of *custom* translation functions", and mention that the other fields follow the default translation behaviour. I didn't quite grok that at the beginning. Or maybe better would be a small docstring at the top of the document explaining how to use the API in two or three sentences.